File: src/interaction/VSphere.py
from espresso import pmi, infinity
from espresso.esutil import *

from espresso.interaction.Potential import *
from espresso.interaction.Interaction import *
from _espresso import interaction_VSphere, interaction_FixedSingleListVSphere

class VSphereLocal(PotentialLocal, interaction_VSphere):
    'The (local) VSphere potential.'
    def __init__(self, a1=1.0, a2=0.0, Nb=1, 
                 cutoff=infinity, shift=0.0):
        """Initialize the local VSphere object."""
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            if shift == "auto":
                cxxinit(self, interaction_VSphere, a1, a2, Nb, cutoff)
            else:
                cxxinit(self, interaction_VSphere, a1, a2, Nb, cutoff, shift)

class FixedSingleListVSphereLocal(InteractionLocal, interaction_FixedSingleListVSphere):
    'The (local) VSphere interaction using FixedSingle lists.'
    def __init__(self, system, vl, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_FixedSingleListVSphere, system, vl, potential)

    def setPotential(self, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, potential)

    def getPotential(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getPotential(self)

    def setFixedSingleList(self, fixedsinglelist):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setFixedSingleList(self, fixedsinglelist)

    def getFixedSingleList(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getFixedSingleList(self)

if pmi.isController:
    class VSphere(Potential):
        'The VSphere potential.'
        pmiproxydefs = dict(
            cls = 'espresso.interaction.VSphereLocal',
            pmiproperty = ['a1', 'a2', 'Nb']
            )

    class FixedSingleListVSphere(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espresso.interaction.FixedSingleListVSphereLocal',
            pmicall = ['setPotential','getPotential','setFixedSingleList', 'getFixedSingleList']
            )

File: src/interaction/VSphere.cpp
#include "python.hpp"
#include "VSphere.hpp"
#include "FixedSingleListInteractionTemplate.hpp"

namespace espresso {
  namespace interaction {
      typedef class FixedSingleListInteractionTemplate< VSphere >
      FixedSingleListVSphere;

    //////////////////////////////////////////////////
    // REGISTRATION WITH PYTHON
    //////////////////////////////////////////////////
    void 
    VSphere::registerPython() {
      using namespace espresso::python;

      class_< VSphere, bases< Potential > >("interaction_VSphere",
      init< real, real, int, real >())
	  .def(init< real, real, int, real, real >())
	  .add_property("a1", &VSphere::geta1, &VSphere::seta1)
	  .add_property("a2", &VSphere::geta2, &VSphere::seta2)
	  .add_property("Nb", &VSphere::getNb, &VSphere::setNb)
      ;

      class_< FixedSingleListVSphere, bases< Interaction > >("interaction_FixedSingleListVSphere",
      init< shared_ptr<System>, shared_ptr<FixedSingleList>, shared_ptr<VSphere> >())
      .def("setPotential", &FixedSingleListVSphere::setPotential)
      .def("getPotential", &FixedSingleListVSphere::getPotential)
      .def("setFixedPairList", &FixedSingleListVSphere::setFixedSingleList)
      .def("getFixedPairList", &FixedSingleListVSphere::getFixedSingleList)
      ;
    }

  }
}
